modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java

```diff
@@ -380,35 +426,106 @@ private CompletableFuture<? extends RaftGroupService> startFollowerNode(
+
+                    handlePeersChange(configuration, raftService);
+
+                    return null;
+                })
+                .whenComplete((res, ex) -> {
+                    if (ex != null) {
+                        LOG.error("Error while handling ConfigurationCommitted event", ex);
+                    }
+                });
     }
 
-    private RaftNodeId raftNodeId() {
-        return raftNodeId(new Peer(clusterService.nodeName()));
+    private void updateRaftClientConfigIfEventIsNotStale(CommittedConfiguration configuration, RaftGroupService raftService) {
+        IndexWithTerm newIndexWithTerm = new IndexWithTerm(configuration.index(), configuration.term());
+
+        lastHandledIndexWithTerm.updateAndGet(existingIndexWithTerm -> {
+            if (newIndexWithTerm.compareTo(existingIndexWithTerm) > 0) {
+                LOG.info("Updating raftService config to {}", configuration);
+
+                raftService.updateConfiguration(PeersAndLearners.fromConsistentIds(
+                        Set.copyOf(configuration.peers()),
+                        Set.copyOf(configuration.learners())
+                ));
+
+                return newIndexWithTerm;
+            } else {
+                LOG.info("Skipping update for stale config {}, actual is {}", newIndexWithTerm, existingIndexWithTerm);
+
+                return existingIndexWithTerm;
+            }
+        });
     }
 
-    private static RaftNodeId raftNodeId(Peer localPeer) {
-        return new RaftNodeId(MetastorageGroupId.INSTANCE, localPeer);
+    private void handlePeersChange(CommittedConfiguration configuration, RaftGroupService raftService) {
+        synchronized (peersChangeMutex) {
+            if (peersChangeState == null || configuration.term() <= peersChangeState.termBeforeChange) {
+                return;
+            }
+
+            PeersChangeState currentState = peersChangeState;
+
+            if (thisNodeIsEstablishedAsLonelyLeader(configuration)) {
+                LOG.info("Lonely leader has been established, changing voting set to target set: {}", currentState.targetPeers);
+
+                PeersAndLearners newConfig = PeersAndLearners.fromConsistentIds(currentState.targetPeers);
+                raftService.changePeersAndLearners(newConfig, configuration.term())
+                        .whenComplete((res, ex) -> {
+                            if (ex != null) {
+                                LOG.error("Error while changing voting set to {}", ex, currentState.targetPeers);
+                            } else {
+                                LOG.info("Changed voting set successfully to {}", currentState.targetPeers);
+                            }
+                        });
+            } else if (targetVotingSetIsEstablished(configuration, currentState)) {
+                LOG.info("Target voting set has been established, unpausing secondary duties");
+
+                peersChangeState = null;
```

Comment:
Should we set it to null if "thisNodeIsEstablishedAsLonelyLeader"? When it becomes null in that case?

Reply:
It works like this for the case when the target set has more than 1 node:

1. `becomeLonelyLeader()` is called, it creates `peersChangeState` and does `resetPeers()`, so after some time the appointed node will elect self as leader and commit the new configuration (where the voting set [aka peers] consists of just the appointed node)
2. It will then see that config committed, so first branch of if/elseif will trigger (`thisNodeIsEstablishedAsLonelyLeader()`), it will try to do `changePeersAndLearners()` to switch to the target voting set. At this moment the `peersChangeState` is still needed for the next step
3. As a result of `changePeersAndLearners()` on the previous step, new configuration (having the target set as the voting set) will be committed, it will be noted by the second branch of if/elseif, so it will finish the transition and after that it will set the state to null